drivers/adc/adc_nrfx_saadc.c

```diff
@@ -521,99 +590,45 @@ static int start_read(const struct device *dev,
-			 * negative input tied to ground. We can therefore
-			 * observe negative values if the positive input falls
-			 * below ground. If the sample bitsize is larger than
-			 * the resolution, we can detect negative values and
-			 * correct them to 0 after the sequencen has ended.
-			 */
-			if ((m_data.single_ended_channels & BIT(channel_id)) &&
-			    (NRF_SAADC_8BIT_SAMPLE_WIDTH == 8 && resolution == 8)) {
-				LOG_ERR("Channel %u invalid single ended resolution",
-					channel_id);
-				return -EINVAL;
-			}
-			/* When oversampling is used, the burst mode needs to
-			 * be activated. Unfortunately, this mode cannot be
-			 * activated permanently in the channel setup, because
-			 * then the multiple channel sampling fails (the END
-			 * event is not generated) after switching to a single
-			 * channel sampling and back. Thus, when oversampling
-			 * is not used (hence, the multiple channel sampling is
-			 * possible), the burst mode have to be deactivated.
-			 */
-			nrf_saadc_burst_set(NRF_SAADC, channel_id,
-				(sequence->oversampling != 0U ?
-					NRF_SAADC_BURST_ENABLED :
-					NRF_SAADC_BURST_DISABLED));
-			nrf_saadc_channel_pos_input_set(
-				NRF_SAADC,
-				channel_id,
-#if NRF_SAADC_HAS_AIN_AS_PIN
-				saadc_psels[m_data.positive_inputs[channel_id]]
-#else
-				m_data.positive_inputs[channel_id]
-#endif
-				);
-			++active_channels;
-		} else {
-			nrf_saadc_burst_set(
-				NRF_SAADC,
-				channel_id,
-				NRF_SAADC_BURST_DISABLED);
-			nrf_saadc_channel_pos_input_set(
-				NRF_SAADC,
-				channel_id,
-				NRF_SAADC_INPUT_DISABLED);
-		}
-	} while (++channel_id < SAADC_CH_NUM);
+	channels_cnt = POPCOUNT(nrfx_saadc_channels_configured_get() & selected_channels);
+	if (channels_cnt == 0) {
+		LOG_ERR("No channel configured");
+		return -EINVAL;
+	}
 
-	error = set_resolution(sequence);
+	error = get_resolution(sequence, &resolution);
 	if (error) {
 		return error;
 	}
 
-	error = set_oversampling(sequence, active_channels);
+	error = get_oversampling(sequence, channels_cnt, &oversampling);
 	if (error) {
-		return error;
+		return -EINVAL;
 	}
 
-	error = check_buffer_size(sequence, active_channels);
+	nrfx_err = nrfx_saadc_simple_mode_set(selected_channels,
+					      resolution,
+					      oversampling,
+					      event_handler);
+	if (nrfx_err != NRFX_SUCCESS) {
+		return -EINVAL;
+	}
+
+	error = check_buffer_size(sequence, channels_cnt);
 	if (error) {
 		return error;
 	}
 
 #if defined(ADC_BUFFER_IN_RAM)
 	m_data.user_buffer = sequence->buffer;
-	m_data.active_channels = active_channels;
+	m_data.channels_cnt = channels_cnt;
 
-	nrf_saadc_buffer_init(NRF_SAADC,
-			      (nrf_saadc_value_t *)m_data.samples_buffer,
-			      active_channels);
+	nrfx_saadc_buffer_set(m_data.samples_buffer, channels_cnt);
 #else
-	nrf_saadc_buffer_init(NRF_SAADC,
-			      (nrf_saadc_value_t *)sequence->buffer,
-			      active_channels);
+	/* Function needs buffer pointer and number of samples not buffer
```

Comment:
this comment does not exactly corresponds to what the function below does, so it might be confusing (it is at least for me).

What about:
```
	/* Buffer is filled in chunks, each chunk composed of number of samples
         * equal to number of active channels. Buffer pointer is advanced and reloaded after each chunk.
         */

```